app/src/org/commcare/android/util/FormUploadUtil.java

```diff
@@ -62,262 +70,309 @@ public static Cipher getDecryptCipher(SecretKeySpec key) {
         }
-        
-        // mime post
-        MultipartEntity entity = new DataSubmissionEntity(myListener, submissionNumber);
-        
-        for (int j = 0; j < files.length; j++) {
-            File f = files[j];
+        return bytes;
+    }
+
+    /**
+     * Add files of supported type to the multipart entity, encrypting xml
+     * files.
+     *
+     * @param entity Add files to this
+     * @param key    Used to encrypt xml files
+     * @param files  The files to be added to the entity,
+     * @return false if invalid xml files are found; otherwise true.
+     * @throws FileNotFoundException Is raised when an xml doesn't exist on the
+     *                               file-system
+     */
+    private static boolean buildMultipartEntity(MultipartEntity entity,
+                                                SecretKeySpec key,
+                                                File[] files)
+            throws FileNotFoundException {
+        for (File f : files) {
             ContentBody fb;
-            
+
             boolean supported = false;
-            for(String ext : SUPPORTED_FILE_EXTS) {
-                if(f.getName().endsWith(ext)) {
+            for (String ext : SUPPORTED_FILE_EXTS) {
+                if (f.getName().endsWith(ext)) {
                     supported = true;
                     break;
                 }
             }
-            
-            //TODO: Match this with some reasonable library, rather than silly file lines
+
             if (f.getName().endsWith(".xml")) {
-                
-                //fb = new InputStreamBody(new CipherInputStream(new FileInputStream(f), getDecryptCipher(aesKey)), "text/xml", f.getName());
-                
-                if(key != null){
-                    if(!validateSubmissionFile(f, FormUploadUtil.getDecryptCipher(key))) {
-                        return RECORD_FAILURE;
+                if (key != null) {
+                    if (!validateSubmissionFile(f)) {
+                        return false;
                     }
-                    fb = new EncryptedFileBody(f, FormUploadUtil.getDecryptCipher(key), "text/xml");
-                }
-                else{
+                    fb = new EncryptedFileBody(f, FormUploadUtil.getDecryptCipher(key),
+                            "text/xml");
+                } else {
                     fb = new FileBody(f, "text/xml");
                 }
-                
                 entity.addPart("xml_submission_file", fb);
-                
-                //fb = new FileBody(f, "text/xml");
-                //Don't know if we can ask for the content length on the input stream, so skip it.
-//                if (fb.getContentLength() <= MAX_BYTES) {
-//                    Log.i(t, "added xml file " + f.getName());
-//                } else {
-//                    Log.i(t, "file " + f.getName() + " is too big");
-//                }
             } else if (f.getName().endsWith(".jpg")) {
```

Comment:
This is a good point. Since most of those conditional bodies are pretty much identical, I was thinking of taking out the entire if/else mess and add some smarter logging mechanism.

Let's leave this for a future refactor, since it we can probably get even more aggressive about the refactor.